### ISSUE TITLE
Metadata Issue

### ISSUE THREAD

**Issue by SkillenUK** [savonet/liquidsoap#384]
First off, I would like to say that I love Liquidsoap and I think that it's an excellent project, a big thank you to everyone involved!

So I have a fairly basic configuration running and my radioref.liq configuration is as follows (I've made some values generic for obvious reasons)...

```
# Enable harbor for any external connection
set("harbor.bind_addr","0.0.0.0")

# DJ live input using direct connection to liquidsoap
djlive = input.harbor("live",port=8010,password="********")

# Auto dj playlist input
autodj = playlist(mode='randomize',"/etc/liquidsoap/autodj/autodj.m3u")

# Set the fallback
input = mksafe(fallback(track_sensitive=false, [djlive,autodj]))

# First transcoder: MP3 64 kbps mono
output.icecast(
  %mp3(bitrate=64, stereo=false),
  mount="/stream64",
  host="localhost", port=8000, password="********",
  name="My Name", description="64kbps stream", url="https://www.mywebsite.com", genre="Dance Music",
  mean(input))

# Second transcoder : MP3 128 kbps stereo
output.icecast(
  %mp3(bitrate=128), 
  mount="/stream128",
  host="localhost", port=8000, password="********",
  name="My Name", description="128kbps stream", url="https://www.mywebsite.com", genre="Dance Music",
  input)
```

I'm not sure if my issue is a bug or if it is a problem with my configuration but hopefully someone will be able to advise me on the best solution either way. When a DJ connects to the input harbor, my playlist is interrupted as expected and the live DJ stream can be heard. My issue is that periodically, the live stream can drop out for a few seconds due to some interruption from the streaming client (for example the internet connection that the DJ has might not be very stable). So because of the interruption, the stream falls back to the playlist, then goes back to the live DJ stream a few seconds later once it has been restored. My issue is that when this happens, the metadata is updated to whatever MP3 is playing from the playlist then when the live DJ stream is restored it does not change back to the metadata for the live stream. So after a drop out of the live stream the metadata shows the incorrect information.

My website has some JavaScript which fetches the metadata from the Icecast server (using json) to display what is currently playing so I would really like to find a solution to this, when it happens the radioref.log shows...

2017/01/14 12:12:15 [decoder:3] Method "MAD" accepted "audio/mpeg".
2017/01/14 12:12:15 [threads:3] Created thread "harbor source feeding" (2 total).
2017/01/14 12:12:15 [harbor_4961:3] Decoding...
2017/01/14 12:12:15 [harbor_4961:3] New metadata chunk ? -- ?.
2017/01/14 12:12:15 [fallback_4965:3] Switch to harbor_4961 with transition.
2017/01/14 12:32:18 [fallback_4965:3] Switch to autodj(dot)m3u with forgetful transition.
2017/01/14 12:33:50 [fallback_4965:3] Switch to harbor_4961 with transition.

So I wanted to report the bug (if it is a bug) but I would also like to find a work around. Does anyone know if there is a feature that would act in the same way as the mksafe() command but with a timeout? For example, I could use it on the live stream so that if the DJ drops out for a few seconds then it will not fall back to the playlist immediately but wait until a timeout has been reached (so the listeners will just hear a few seconds silence instead of a few seconds of another song during the interruption and the metadata isn't changed). Another option I'm considering is adding a large buffer to the live stream (10 seconds or so) but I'm not sure if this will cause other issues. Thanks in advance for any help on this, regards,

Chris.

**Comment by desavil** [comment]
I attaching to your request - mksafe or some buffor option for Live Harbor.

**Comment by toots** [comment]
Yeah, the problem is that connection dropping is considered an end-of-track thus getting rid of last metadata which would otherwise be replayed. I've pushed PR #385 to try to fix this. I think I'm gonna need @dbaelde's input though before merging.

**Comment by SkillenUK** [comment]
Thanks for the quick response on this guys! I've increased the log level to 4 and done a bit more testing with the DJ who is having the issue so if you want any additional logs from me then let me know and I can provide them, regards,

Chris.

**Comment by toots** [comment]
Ok, should be closed now, mind trying the latest git code?

**Comment by SkillenUK** [comment]
I'm more than happy to test but I currently have it installed from the Ubuntu repository. I’m guessing that I can download and make / build it to another folder then run from here for testing? Obviously I will stop the existing service first. Out of interest, how often will the build in the Ubuntu repo be updated? Thanks,

Chris.

**Comment by toots** [comment]
Hi,

We are not maintaining the ubuntu/debian builds anymore as we've switched to opam. Installing the latest code with opam should be fairly easy tho, see: http://liquidsoap.fm/download.html, OPAM section and then Latest development version section.

Once you have a build of the recent code, make sure that you enable the `replay_metadata` option on `input.harbor`: `input.harbor(replay_metadata=true,...)`.

Thanks!

**Comment by SkillenUK** [comment]
Hi,
I attempted to remove the Ubuntu package and install using this method but I got in to a bit of a mess with it. Maybe I hadn't completely removed everything (I'm not an expert with this but I am learning!). Luckily I've managed to roll back to the Ubuntu package and after a bit of playing get it running again. Bare with me, I'll set up a VM later this week and try to work it out on there. If I'm successful then I'll try again on my main server, thanks,

Chris.

**Comment by toots** [comment]
Hey,

Yeah, the potential fix was merged and released so hopefully your issue should be fixed with the latest `opam`.

Of course, feel free to send me your remarks/notes on opam install. I'm at: toots@rastageeks.org

**Comment by SkillenUK** [comment]
Great, thanks I'll test it.

I've made notes on how I installed it with OPAM in Ubuntu so this could be useful to others who might be struggling. I wouldn't call myself a beginner in Ubuntu but I'm also not an expert, I usually get there in the end! I found it tricky to work out how to install with OPAM under a generic user that doesn't have elevated privileges (obviously OPAM doesn't allow installing under the sudo user). It was also tricky getting all the required dependencies and as this is the first time I've used OPAM there was some learning to do on my side. I'll try and put something together and send it to you :)
